Allow customizing repository used for AMI configuration

elad opened the pull request:
This is a reference pull request related to the issues raised in #72, specifically allowing a custom repository (instead of the one used to create the AMI) and allowing more than one hard-coded key for commit verification.

elad committed:
Initial take at custom repository based configuration.

mlococo commented on README.md:
```diff
@@ -80,6 +80,14 @@ Options
         Allows you to use your own reflector
         Default: http://reflector2.datastax.com/reflector2.php
 
+    --repository <repository>:<branch>
```
Standard separator to split on seems to be '/' instead of ':'. This is what hub (https://hub.github.com/) and bundler (http://hashnuke.com/2012/06/05/bundler-new-github-option.html)

I can fix this one post-merge.

mlococo replied:
I think this might need to be 2 options.  A repo can look like a wide variety of things including several URL formats and filesystem paths (GIT_URL and REMOTES sections of https://www.kernel.org/pub/software/scm/git/docs/git-fetch.html). Are we sure there's a character that's safe to split on? Colon is certainly not safe (shows up in all the URL formats). I'd say leave the repo by itself in this option.

I think we can use the existing --force_commit for the branch-reference.

elad replied:
Agree, done.

mlococo commented on ds0_utils.py:
```diff
@@ -75,3 +81,38 @@ def required_commit():
+    else:
+        default_key = {
+            'id': '7123CDFD',
+            'file': '/home/ubuntu/datastax_ami/repo_keys/DataStax_AMI.7123CDFD.key',
```
@joaquincasares Is this path reasonable to hardcode or do we have some config-item we should key this basepath off of?

elad replied:
This reminds me: Naming files in `repo_keys` as `<key_id>.key` will make them easier to work with from scripts.

mlococo replied:
I have no strong opinion here. It's easier to glance at the directory and know what the keys are for they have human readable names, but easier to reference them unambiguously from code if they're named after their id's.

Unless @joaquincasares has a strong opinions, I'll merge it as is, or if you want to fix the key-names I'm fine with that as well.

elad replied:
Feel free to merge it as-is. I can always submit a pull request later if @joaquincasares has no objection.

joaquincasares replied:
Using a hardcoded path is what I've been doing so far and would ideally be accepted here.

I chose the repo key naming convention to be able to easily jog my memory as to why we store certain keys as well as documenting the keys for the community. If neither of you two think there's an issue with changing the naming format, I'd be open to the change.

elad replied:
First, clarification - when I said "feel free to merge it as-is" I was referring to the key file naming convention, not the hard-coded path. :)

Since the primary consumer of the key files is a program, I think it makes sense for the naming convention to be easily processed with one. I'll submit a separate pull request that addresses this issue once we're happy with these changes.

mlococo commented on ds0_utils.py:
```diff
@@ -75,3 +81,38 @@ def required_commit():
+    options = parse_ec2_userdata()
+
+    if options and options.repository:
+        if ':' in options.repository:
```
Note to self: split on '/' instead of ':'

elad committed:
Let --repository flag specify only the repository

Branch/tag can be specified using --forcecommit, as noted by @mlococo.

While here, fix comment.

elad committed:
Update README.md.

mlococo commented on ds0_updater.py:
```diff
@@ -5,13 +5,36 @@
 
     # update the repo
-    logger.exe('git pull')
+    logger.exe('git pull ' + ds0_utils.repository())
```
Unless we do some rejiggering from line 37 to 44, we have something less flexible than what you started with. I'm actually fine with this, though... it means:
1. --force_commit only works for commits that are in the official riptano/ComboAMI repository... and so is mainly useful for project commiters to test things.
2. --repository fast-forwards to the latest commit in the 2.5 branch of a user-specified repo. It's mainly useful to community members that want to tweak the AMI startup-scripts in a lightweight way. They can't specify an arbitrary commit or branch, and must push their changes to a fork with a branch named '2.5'.

While this isn't arbitrarily flexible, it seems to me that it's still quite useful. I'm fine with this unless you want to rethink the git strategy. I don't have a better suggestion off the top of my head.

elad replied:
Ah, you're right - I forgot git would default to the branch if none was specified. This is my mistake and wasn't my intention. In that case I would like to add a `--branch` flag to allow specifying both repository and branch, otherwise users can't keep more than one configuration in a repository.

mlococo replied:
Hold off on that for the minute. I'd like to ponder this overnight. I think we can do this without additional options, and possibly with fewer potentially conflicting changes to the git workspace.

Totally unfiltered brainstorming:
1. Maybe git-remote-add instead of git-pull? This would pull in the commit-history of the new remote and let us use any refspec in it.
2. Or can git-fetch be used without first adding a new remote to allow references to remote refspecs? Probably not.
3. Or can git-pull take any refspec and not just a branch name? If that's the case we can just tack forcecommit onto the git-pull command and be done (maybe changing "origin" to be the default repo instead of "").

elad replied:
My thinking was that it makes sense to stick to elements that users of git (and I assume most version control systems) are familiar with: repository, branch, tag, and commit id. So the interface (options passed in user data) should take one or more of them.

If we want to stick to one option, maybe another direction to consider is [what `npm` uses](https://docs.npmjs.com/files/package.json#git-urls-as-dependencies) for dependencies. It could have a default like:

```
--repository https://github.com/riptano/ComboAMI#2.5
```

When used to replace `--forcecommit`:

```
--repository https://github.com/riptano/ComboAMI#33f501117ba3c4238c28606463315063b5afb68d
```

When used to point to a custom repository:

```
--repository https://github.com/elad/ComboAMI#my_branch_or_tag
```

I would actually prefer it, especially given there's prior art and logic we can borrow for the code. What do you think?

mlococo replied:
Using the hash as a repo/refspec separator looks fine to me. If npm is doing it, it's probably sensible. I also agree it's a much more intuitive interface than the --forcecommit flag.

It would be very nice if --repository could accept any refspec. Right now, it can only accept a branch name because git doesn't know the commit-hashes for the custom-repo. Rather than git pulling (which cannot accept a commit-hash from the remote) Can we git-remote-add (which will grab the commit-history with hashes), followed by a git reset --hard <refspec> to modify the workspace which can accept a branch, tag, or commit-hash from either repo. I'm fine hard-coding the name of the remote since we can only support a single one anyway.

It would also be good for --forcecommit and the refspec piece of the --repository option to be synonyms. They do the same thing, which is reset the workingspace to a particular refspec. I think we don't want to drop --forcecommit for backwards compatibility reasons, but could the repository function:
1. Return the offical repo by default
2. Check both --forcecommit and the refspec piece of --repository, return whichever is set, and if both are set return the refspec set in --repository and log a warning.

With these changes, I'm +1 on this PR and will mege/test it. Thanks a ton for doing this and for being so responsive to feedback.

elad replied:
> It would be very nice if --repository could accept any refspec.

I agree. Since I don't think `npm` supports that though, we'll have to come up with syntax. How about "if it starts with `+` it's a refspec, otherwise it's an `npm`-compatible URL"?

As for implementation, I like your idea. Let's think (out loud) about implementation. We break down whatever we get to repository, branch/tag, and commit-id. We have defaults - `https://github.com/riptano/ComboAMI`, branch `2.5`, and none (=latest), respectively. We can then do something like:

```
git remote remove origin
git remote add origin <repository>
git pull
git checkout <branch/tag> or git checkout tags/<branch/tag> if unsuccessful
git reset --hard <commit-id>
```

I think that lets us maintain backwards compatibility for `--forcecommit` by relying on defaults.

Does that make sense? Am I missing something?

> Thanks a ton for doing this and for being so responsive to feedback.

Sure! :) My pleasure. You guys are responsible for fantastic software and I'm sure we're not the only one this AMI is saving a lot of time for.

mlococo replied:
Ugh... I just realized I've been using "refspec" incorrectly my whole life. Anyplace I've used "refspec" previously, I really meant something like commit-reference as defined here: http://git-scm.com/book/en/v2/Git-Tools-Revision-Selection... so branch, short-sha, long-sha, reflog-references, etc.

So we get that for free with the following implementation of git commands:
git remote remove origin
git remote add origin <repository>
git fetch # Use fetch instead of pull because it leaves the workspace untouched
git reset --hard <commit> # You can reset to an arbitrary commit without checking out

In the above, all of the following are valid commit references:
- origin/dev-2.5
- 0d0d8b4
- Several other formats

The only caveat here is that branch-names need to be prefixed with the remote name, which can be documented and hardcoded to origin. Or perhaps leave origin in place and hardcode name for the remote of the custom-repo to somthing like 'custom-repo'.

I'm not in love with the fiddliness of the hard-coded remote name bubbling up in the user-interface, but this can reference any branch-name or commit-id, has only a single workspace transformation, and makes --forcecommit/second-half-of-repository to be a synonym while maintaining backward compat. All of which feels pretty clean.

I've had to do some git research during this, so if you think I'm missing something that makes it bad/unworkable, call it out.

elad replied:
> The only caveat here is that branch-names need to be prefixed with the remote name, which can be documented and hardcoded to origin. Or perhaps leave origin in place and hardcode name for the remote of the custom-repo to somthing like 'custom-repo'.
> 
> I'm not in love with the fiddliness of the hard-coded remote name bubbling up in the user-interface, but this can reference any branch-name or commit-id, has only a single workspace transformation, and makes --forcecommit/second-half-of-repository to be a synonym while maintaining backward compat. All of which feels pretty clean.

Doesn't the assumption that the remote name is origin relieve us from bubbling it up to the user interface? After all, wherever the user would put "origin" we can have them omit it and add it implicitly.

Otherwise it seems to be the right thing to do - of course we'll run tests to make sure it all works :) - but I think we're on the right track here.

mlococo replied:
With so many possible formats for that could come into the second-half of the --repository param, I'm not sure how you could easily figure out when to prepend origin or not (can you do that for a commit-hash, for example?), unless git let's you be implicit with origin, but I'm pretty sure I tested and it didn't. But yeah, if you can make it work implicitly that would be sweet.

elad replied:
Why would the remote name bubble up to the user interface? IIUC, the remote name is what the repository is referred to as locally and isn't exposed when sharing URLs. So it's unclear to me when would `origin` (or any other name) be part of anything that one could pass to `--repository`.

I'm getting a bit confused so let's stick to examples, that is, closer to implementation. :) I assume our starting condition is a cloned official repository:

```
git clone https://github.com/riptano/ComboAMI
```

The default branch will be whatever the default is, right now `2.5`. So after the initial clone `git status` should print this:

```
$ git status
On branch 2.5
Your branch is up-to-date with 'origin/2.5'.

nothing to commit, working directory clean
$
```

At that point our bootstrapping script (`ds0_updater.py`) kicks in, parses the `--repository` argument (and sets defaults), and executes a bunch of git commands.

The `--repository` argument is a subset of what `npm` supports: a URL optionally followed by a `#` followed by a commit-ish which is "any tag, sha, or branch which can be supplied as an argument to `git checkout`." Examples:

```
--repository https://github.com/elad/ComboAMI
--repository https://github.com/elad/ComboAMI#my_tag
--repository https://github.com/elad/ComboAMI#my_branch
--repository https://github.com/elad/ComboAMI#e5e3d41fb5f12461509aa1b6079413b381930d81
```

The defaults can either be hard-coded or inferred from the initially cloned repository (easier maintenance I think). In our case they'd be:

```
repository = 'https://github.com/riptano/ComboAMI'
branch_or_tag = '2.5'
commit_id = None
```

The first git commands executed by `ds0_updater.py` reset the origin:

```
git remote remove origin
git remote add origin <repository>
git fetch
```

Now - if I understand correctly - we want to use `git reset --hard` to switch to the desired commit rather than `git pull` or `git checkout`. So first we translate our `branch_or_tag` to `commit_id` if the latter is still unset using `git show-ref`:

```
pattern = 'refs/remotes/origin/' + branch_or_tag + '|' + 'refs/tags/' + branch_or_tag
commit_id = git show-ref | grep -E <pattern> | cut -d' ' -f1
```

Finally we modify the workspace:

```
git reset --hard <commit_id>
```

That should work, but has one drawback: while we will have the correct revision, we won't _technically_ be on the branch/tag the user specified. Personally I don't see much of a problem in that, even more so if it's documented.

Does the above match what you had in mind? It seems to work for me without exposing that `origin` is the remote name outside the internal scripts, but again I might be forgetting something. :)

mlococo replied:
What I was getting at with leaking the remote ("origin") into the second-half of the --repository flag is this:

```
% git remote # No remote repositories defined

% git branch -a # No branches from any remotes checked out locally
* scratch-branch

% git remote add origin git@github.com:riptano/ComboAMI.git                                                
% git fetch                                                                                                !10195
From github.com:riptano/ComboAMI
 * [new branch]      2.0        -> origin/2.0
 * [new branch]      2.1        -> origin/2.1
 * [new branch]      2.2        -> origin/2.2
 * [new branch]      2.3        -> origin/2.3
 * [new branch]      2.4        -> origin/2.4
 * [new branch]      2.5        -> origin/2.5
 * [new branch]      dev        -> origin/dev
 * [new branch]      dev-2.5    -> origin/dev-2.5

% git reset --hard 2.5 # Resetting to a bare-branch name doesn't work
fatal: ambiguous argument '2.5': unknown revision or path not in the working tree.
Use '--' to separate paths from revisions, like this:
'git <command> [<revision>...] -- [<file>...]'

% git reset --hard origin/2.5 # But resetting to a remote/branch combo does
HEAD is now at e5e3d41 Merge pull request #71 from riptano/dev-2.5

% git reset --hard 33f5011 # As does resetting to a commit-hash
HEAD is now at 33f5011 Reorder chowing
```

A few other useful commit-reference formats can be passed directly to git-reset as well. So without any custom-code to detect or lookup branch-names vs commit-hashes, both work in the above scheme (with the drawback that branch-names must be prefixed by a remote-name). You may be able to improve upon this with some combination of checkout and pull but this is quite simple to implement and does quite a lot.

At this point I'd say use your judgement, you've got a firm understanding of the issues in play and given the code and analysis you've submitted so far I'm sure whatever you come up with will be mergeable.

elad replied:
> `% git reset --hard origin/2.5 # But resetting to a remote/branch combo does`

Ah - I didn't know that, okay. I'll implement and test the changes as soon as I can and push for review.

elad committed:
Change implementation based on discussion with @mlococo.

elad committed:
Update README.md.

elad commented:
Hey, just pushed a bunch of changes that I think address the points raised in our discussion. I'm using `git rev-parse` to detect if the second half of the argument to `--repository` is a commit id or not. In the latter case I prefix it with the remote name (origin) so it doesn't bubble to the user interface. Backwards compatibility is maintained and if `--forcecommit` is specified it takes precedence over everything else and forces use of the official repository.

I hope to test it later today.

mlococo commented:
Awesome, I didn't even know rev-parse existed. I'll have a gander to review
later today.

On Tue, Apr 14, 2015 at 8:55 AM, Elad Efrat notifications@github.com
wrote:

> Hey, just pushed a bunch of changes that I think address the points raised
> in our discussion. I'm using git rev-parse to detect if the second half
> of the argument to --repository is a commit id or not. In the latter case
> I prefix it with the remote name (origin) so it doesn't bubble to the user
> interface. Backwards compatibility is maintained and if --forcecommit is
> specified it takes precedence over everything else and forces use of the
> official repository.
> 
> I hope to test it later today.
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/riptano/ComboAMI/pull/74#issuecomment-92809671.

elad committed:
Two fixes to make this work.

- It's "git remote rm" and not "git remote remove"
- We'll add --allowed-keys handling in a separate pull request

elad commented:
Okay, this seems to work well. I tested by deploying the AMI, wiping the configuration and resetting the repository to my branch, then running `ds0_updater.py`. I'm sure you guys have a better testing environment - please give it a spin. :)

joaquincasares commented:
Hopping in here for just a bit. :)

Unfortunately that would be the best way to test this pull request with an "official test" happening at bake time.

@mlococo If launching a normal AMI with the --dev parameter once this has been merged into dev-2.5 works without complications, then I'd typically merge the pull request into 2.5. That way we ensure we don't break the current AMI and can always retest functionality after the next bake.

The reason for this is that ds0 and ds0_util should never be reloaded after instantiation to ensure the security check is working as expected.

elad commented:
@joaquincasares since you're already here and mentioned rebaking the AMI, I should note that there's at least one more change I'd like to propose in a separate pull request: Adding `--allowed-keys` and letting folks other sign repositories. I think this is obvious given `--repository` and there's an existing issue for it (#69).

joaquincasares commented:
+1, I'd be open for that pull request. I'll leave that for @mlococo to decide if we can get that implemented before the next rebake.

mlococo commented:
Reviewed code, looks awesome. Merged to dev-2.5, I won't have time to test it today as I'm slammed with other work, but will do so this week.

elad commented:
Fantastic, thanks! :)
